Add docs for using feature flags in govuk-frontend (#46)
alphagov/design-system-team-docs: source/development/feature-flags.html.md

---
title: Feature flags in GOV.UK Frontend
weight: 1
last_reviewed_on: 2023-11-28
review_in: 6 months
---
# Using feature flags in GOV.UK Frontend

> Feature flags are a way for us to release large scale or experimental work early by allowing users to opt in our out of a new, breaking change or feature. Use this guidance to decide if a feature flag is useful for your use case.

Reviewer: ```suggestion
Feature flags are a way for us to do a 'soft' release of large scale or experimental work by allowing users to opt in our out of a new, breaking change or feature. Read this guidance to decide whether a feature flag is a useful addition to your project.
```

Reviewer: Experimenting with a way to remove 'early' which felt a bit orphaned within the sentence and some repetition of use/useful in the second sentence. Not essential, so it's a decision as to whether you think it flows better or not.

Author: I like it

## When a feature flag is useful

> Feature flags are useful when the following conditions are true.

Reviewer: It could be because I'm reviewing in this layout, but this sentence looks out of place on its own. Is it worth having the subheads in a bullet list here as well? As a kind of summary?

Author: I think that's a good idea. I've implemented this, let me know what you think.

### The thing is a breaking change

> Feature flags should only be used for things considered breaking changes. This is something where users will potentially need to change their service, or assess if their service needs changing, in order to continue staying up to date with govuk-frontend. In this case, a feature flag would be a way for a user to stay up to date whilst also opting out of the feature for a period of time. We don’t need to do this for minor releases or new features such as a new component because a user can opt out by simply not using the new feature.

Reviewer: ```suggestion
You should only use Feature flags for things considered breaking changes. Assess whether a breaking change will cause a need for users to amend their service in order to stay up to date with govuk-frontend. In this case, a feature flag would be a way for users to stay up to date while also being able to opt out of the feature for a period of time. Feature flags are not needed for minor releases or new features, such as a new component, because it is simpler to opt out by not using the new feature.
```

Reviewer: I think I've changed the meaning of this. I was trying to clarify who was doing what action.

Author: I think the only issue with the suggestion is the 2nd sentence. I was trying to clarify what a breaking change is to make it clear why a feature flag is only necessary for breaking changes. I do that at the end with the new feature explanation though. What do you think about me just removing that 2nd sentence?

Reviewer: I see. I've nicked the breaking change definition from the frontend site as a replacement in the second line. Does it work in this context?

Author: It does! Accepting.

> ### The thing might be a long term project for users

Author: "The feature"? "The change"?

> ### The thing might be a long term project for users

Reviewer: 'feature' risks getting confused with specific 'new features'. 'change' might work. Or perhaps 'action'?
or I'm probably overthinking it and it doesn't really matter.

> ### The thing might be a long term project for users

Reviewer: Let's give it whirl and see what others think

> ### The thing might be a long term project for users

Author: I tried replacing everything with action and it didn't feel right in a few places so I've used a mix of "action" and "change" depending on the context. What do you think?

> ### The thing might be a long term project for users

Reviewer: Sounds like a good plan

> When assessing if a breaking change should be behind a feature flag, consider the scope of work required by the user. A feature flag is useful if the change is very difficult to implement and/or has wide scope. Using the green link example in the previous section, this would impact all links in a given service so that service will likely need to check all their pages to ensure the design of their pages still made sense against the new link colour.

Reviewer: I can't find the green link example. Has it changed name during editing?

Author: Oops I removed it, good spot. I'll put it back in and try to make it more concise.

Author: Done. Whatcha reckon?

Feature flags are more useful for wide scope changes that are likely to take a long time to action over difficult to implement changes. For difficult to implement but short term changes, consider using a pre-release instead (see below section on pre-releases).

### We are unsure of the thing’s value

> This is an alternative to the previous principal on implementation difficulty where, for whatever reason, we have a significant change that we aren’t 100% confident in and have been unable to verify its value. In this instance, putting it behind a feature flag would allow us to get early feedback from users who’re willing to use and test our new/updated feature and give us feedback, allowing us to iterate the change with data.

Reviewer: ```suggestion
This is an alternative to the previous principle on implementation difficulty, where we are not 100% confident of the value of a significant change and need to test it further. In this instance, putting our new/updated feature behind a Feature flag would allow us to get early feedback from users willing to use and test it, which enables us to iterate the change with data.
```

## When a feature flag isn't useful

> Even if you have a scenario where feature flags may be useful, consider the following before implementing one.

Reviewer: ```suggestion
Even if you have a scenario where feature flags might be useful, consider the following before implementing one.
```

### When releasing the thing carries no significant risk

> We should avoid using feature flags in most cases. We have alternative systems for communicating big or impactful changes such as feature depreciation where we tell people we’ll be removing something in a future breaking release and breaking releases themselves where users know that by upgrading, they might need to change something in their service. Feature flags should be used as a way to alleviate instances where releasing as normal is challenging.

Reviewer: Q: do you want all mentions of the feature flag to be first word capitalised, both words capitalised or all lowercase unless starting a sentence?

Author: I personally think all lowercase. Does that feel sensible?

Reviewer: let's try it. we can always change it later if it doesn't work for some reason.

### We could use a pre-release instead

> Another method we have for testing breaking changes before release are pre-releases. This is a version of govuk-frontend that we release to npm like normal releases, but we tag it as a “beta” release eg: instead of 5.0.0, we might release 5.0.0-beta.0. This communicates that it isn’t the next major version of govuk-frontend but an early, “experimental” version of that release, allowing users to assess it before we release that major version.

Reviewer: ```suggestion
We also use pre-releases to test breaking changes before the full release. A pre-release is a version of govuk-frontend that we release to npm as if its a normal release, but we tag it as a “beta”. For example, instead of the tag 5.0.0, we might use 5.0.0-beta.0. This communicates that it is not the next major version of govuk-frontend, but an early “experimental” version for users to assess with their service.
```

Reviewer: I've tweaked this, but noticed references to 'release', 'normal release', and 'major version'. Do we need to choose just one naming convention for consistency or is it clear enough as is?

Author: This is a tricky one because in the context of this they all mean different things.

- release: the thing we publish every time we want to, erm, release something
- normal release: a differentiator between pre-release and a release that isn't a pre-release
- major version: this could be swapped out on balance for something like "breaking release"

Can I make one small suggestion?

>We also use pre-releases to test breaking changes before **a breaking** release. A pre-release is a version of govuk-frontend that we release to npm as if its a normal release, but we tag it as a “beta”. For example, instead of the tag 5.0.0, we might use 5.0.0-beta.0. This communicates that it is not the next major version of govuk-frontend, but an early “experimental” version for users to assess with their service.

That's probably a personal preference. It just feels a bit nicer to me.

Reviewer: That's cool then as they are all different things. As long as it's clear to the user, it's all good. And your suggestion is a good idea.

> Pre-releases are ideal for short term investigation or early access to the community when a breaking change might include something difficult to implement. They’re also generally easier to manage compared to feature flags. Consider if you could use a pre-release instead of a feature flag.

Reviewer: ```suggestion
Pre-releases are ideal for short term investigations or providing early access to the community so they have time to assess a breaking change that might be difficult to implement. Also, pre-releases are generally easier to manage compared to feature flags. Consider if you could use a pre-release instead of a feature flag.
```

## Types of feature flag

> There are 3 broad types of feature flag. These are defined more by the change you’re making rather than implementation of the flag itself.

Reviewer: ```suggestion
There are 3 general types of feature flag. These are more accurately defined by the change you’re making rather than implementation of the flag itself.
```

Author: Reading this back, I wonder if we should simplify this:

>There are 3 types of feature flag. These are defined by the change you’re making rather than implementation of the flag itself.

What do you think?

Reviewer: As a project manager once said to me - if you can take words out of a sentence and it still means the same thing, use fewer words.

### Experimental

> Early previews of work that will likely go into a future release. Enables us to receive feedback on something we’re not 100% confident in from members of the community who’re willing to test it in their service. Use an experimental flag to release a long term change early to assess its value in a live environment.

Reviewer: ```suggestion
Use a flag to create early previews of a future release to help assess its value in a live environment. In this case the flag will enable you to receive feedback from members of the community who’re willing to test early versions in their service.
```

Reviewer: Some of the content in this doc reads as duplication. Is it worth looking at ways to merge any duplication or is it repeated for contextual reasons?

Author: I'm guessing you're referring to the stuff in this section where I define the different classifications of feature flag, which I then go into again when defining when to use a feature flag at all. Probably we can merge them. Knowing the types of feature flags is potentially useful if we end up using them more so that we can establish common language for our users, however it's not really necessary.

The secret main reason this section exists is to house the "extended deprecation" heading and tell people not to delay deleting a feature unless we absolutely have to. Probably we can convey this somewhere else.

Reviewer: Honestly, I've forgotten what I was specifically referring to. If re-emphasising something is important for the context, then let's not shuffle content around unnecessarily. We can always do some lite testing of how people read/use the content to double-check it's clear enough.

Author: Lets leave it for now and see how we get on

### Extended release

> Where we release a new feature or change but not have it available by default. Allows users to implement a large or complex feature incrementally whilst enabling them to continue staying up to date with new releases of our library. Use an extended release to facilitate the rollout of a long term, wide reaching feature.

Reviewer: ```suggestion
You can use a flag to facilitate an extended release of a new long term, wide reaching feature or change, while not making it available by default. This allows users to implement large or complex features incrementally while enabling them to continue staying up to date with new releases from our library.
```

### Extended deprecation

> A “reverse” extended release where we remove something but keep the feature available via a feature flag. This lets users continue using features we want to remove for an extended period of time, allowing them to move away from said features incrementally. **Only use an extended deprecation as a last resort**. Whilst extended releases enable us to roll out changes slowly, extended deprecations moreso than extended releases facilitate potential bad or outdated practice hanging around. We should feel empowered to remove things that are bad for users.

Reviewer: ```suggestion
You can use a flag as a 'reverse' extended release where you remove a feature, but keep it available through a feature flag. This is a form of extended deprecation which gives users the option to stop using removed features incrementally. 
**Only use an extended deprecation as a last resort**. While extended releases enable you to roll out changes slowly, extended deprecations are more likely to facilitate bad or outdated practices. In the first instance, you should aim to remove features that are outdated or have issues.
```

## Additional considerations

> Once you’ve established that a feature flag is a useful option for the thing you’re working on, consider the following principles when implementing and operating your flag.

Reviewer: ```suggestion
Once you’ve established that a feature flag is a useful option for your project, consider the following principles when implementing and operating your flag.
```

### Try to minimise the scope of the change

> Try to have your flag controlling as little as possible. Minimising how much is behind a feature flag reduces the burden of technical management and the impact of the change itself. Check if there’s anything in your change that can be released outside the feature flag as a separate change.

Reviewer: ```suggestion
Try to have your flag controlling as little as possible. Minimising how much is behind a feature flag reduces the burden of technical management and the impact of the change. Check if there’s anything in your change that you can release outside the feature flag as a separate change.
```

### Be aware of the need for multiple feature flags

> Your feature may impact multiple parts of our tech stack eg: the Sass and the javascript. In which case you should ensure that your flag is built in such a way that the impacted element states are “synced” where possible. For example, if you have a flag in a component macro that controls the markup of that macro but also requires a Sass change, your flag should automatically add a CSS class that applies that styling change.

Reviewer: ```suggestion
Your feature might impact multiple parts of our tech stack, for example, the Sass and the javascript. If this happens, you should make sure your flag is built so the impacted element states are “synced” where possible. For example, if you have a flag in a component macro that controls the markup of that macro, but also requires a Sass change, your flag should automatically add a CSS class that applies that styling change.
```

## The lifecycle of a feature flag

> Follow these instructions for applying, working with and finally getting rid of feature flags.

Reviewer: ```suggestion
Follow these instructions for applying, working with and removing feature flags.
```

### Implementing a feature flag
Feature flags and the change they’re controlling should be released in a minor release aka: X.X.X. As part of this release, automated tests should be written to account for govuk-frontend with the feature flag on and off to check for unforeseen impact to parts of the codebase that the flag wasn’t intended to affect.

> At this stage, we should create github issues to remove the feature flag in a future major release and apply them to ourplanning accordingly.

Reviewer: ```suggestion
At this stage, you should create GitHub issues to remove the feature flag in a future major release and add them to the relevant plans.
```

> We communicate in release notes and guidance what this feature flag impacts, how to use it and how “experimental” the new change is.

Reviewer: ```suggestion
We use release notes and guidance to provide information on any impacts from a feature flag, how to use it and whether the new change is 'experimental'.
```

### Working with a feature flag in flight

> By turning a feature flag on, we presume a user is opting into the following conditions:

Reviewer: ```suggestion
By turning a feature flag on, you presume a user is opting into the following conditions:
```

Reviewer: Should the sentences following this be in bullet points or should there be a summary list of bullet points related to the following sections/subheads?

Author: It should be. I think I planned for it to be a list and then forgot to mark it up. Will sort.

> We may release “breaking” changes to things behind the feature flag. We will communicate this in release notes.

Reviewer: ```suggestion
We might release 'breaking' changes to things behind the feature flag. We will communicate this in release notes.
```

> Depending on the change itself and if it’s experimental or not, the thing behind the feature flag may not be 100% operable ie: there may be bugs, users may provide negative feedback etc.

Reviewer: ```suggestion
Depending on what the feature is and whether it’s experimental or not, the change behind the feature flag might not be 100% operable. For example, there might be bugs or users might provide negative feedback.
```

> It’s important for us to monitor community feedback on the feature flag, its use and the change itself so that we can iterate the change before it’s removal.

Reviewer: ```suggestion
It’s important to monitor community feedback on a feature flag, how it's used and how well the change functions, so we can iterate the change before the flag's removal.
```

### Removing a feature flag

> Feature flags should be removed in major versions of govuk-frontend and their removal should be considered breaking. Ideally we should plan for removal of feature flags in the next major release after the feature flag is implemented eg: if we release a feature flag in 5.2.0, we should remove it in 6.0.0. There are numerous reasons why we might not always be able to do this however, such as users commenting that they need more time whilst we have other pressing breaking changes we want to make. Therefore, if we can’t remove the feature flag in the following major release, we should decide per future breaking release if the feature flag can be removed.

Reviewer: ```suggestion
You should always aim to remove feature flags in the next major version of govuk-frontend after the flag was implemented and consider their removal as a breaking change. For example, if you release a feature flag in 5.2.0, you should remove it in 6.0.0. There are many reasons why this might not always be possible, such as users asking for more time or other breaking changes taking priority. If you cannot complete the removal in the next major release, you should plan to remove the feature flag in a future breaking release.
```

Reviewer: Let me know if you want to meet to speed up the editing process.

Reviewer: How about starting the section with something like:
'You should maintain a consistent process where you check for active feature flags and decide if you still need them.'

Author: Implemented 😎

---
title: Feature flags in GOV.UK Frontend
weight: 1
last_reviewed_on: 2023-11-28
review_in: 6 months
---
# Using feature flags in GOV.UK Frontend

> Feature flags are a way for us to do a 'soft' release of large scale or experimental work by allowing users to opt in our out of a new, breaking change or feature. Read this guidance to decide whether a feature flag is a useful addition to your project.

Reviewer: ```suggestion
Feature flags are a way for us to do a 'soft' release of large scale or experimental work. They allow users to opt in or out of changes that are new, breaking or a feature. Read this guidance to decide whether a feature flag is a useful addition to your project.
```

Author: This slightly changes the meaning of the end of the first sentence, now the second sentence with your suggestion. I was trying to say that a user can now opt out of a feature or change that is new or breaking. Maybe something like...

>They allow users to opt in our out of new or breaking changes.

Thonks?

Reviewer: Yup, ok cool. But should it be 'or' rather than 'our'?
